modules/metastorage-server/src/main/java/org/apache/ignite/internal/metastorage/server/RocksDBKeyValueStorage.java

```diff
@@ -0,0 +1,1249 @@
+
+                            boolean empty = true;
+
+                            if (nativeIterator == null)
+                                nativeIterator = db.newIterator(options);
+
+                            for (nativeIterator.seek(revisionPrefix); nativeIterator.isValid(); nativeIterator.next()) {
+                                empty = false;
+
+                                byte[] key = rocksKeyToBytes(nativeIterator.key());
+
+                                if (p.test(key)) {
+                                    nextRetRev = curRev;
+
+                                    return true;
+                                }
+                            }
+
+                            if (empty) {
+                                nativeIterator.close();
+
+                                nativeIterator = null;
+
+                                return false;
+                            }
+
+                            lastRetRev++;
+                        }
+                    }
+                    finally {
+                        lock.unlock();
+                    }
+                }
+
+                /** {@inheritDoc} */
+                @Override public WatchEvent next() {
+                    Lock lock = rwLock.readLock();
+                    lock.lock();
+                    try {
+                        while (true) {
+                            if (nextRetRev != -1) {
+                                boolean empty = true;
+
+                                List<EntryEvent> evts = new ArrayList<>();
+
+                                for (; nativeIterator.isValid(); nativeIterator.next()) {
+                                    empty = false;
+
+                                    byte[] key = rocksKeyToBytes(nativeIterator.key());
+
+                                    Value val = bytesToValue(nativeIterator.value());
+
+                                    if (p.test(key)) {
+                                        Entry newEntry;
+
+                                        if (val.tombstone())
+                                            newEntry = Entry.tombstone(key, nextRetRev, val.updateCounter());
+                                        else
+                                            newEntry = new Entry(key, val.bytes(), nextRetRev, val.updateCounter());
+
+                                        Entry oldEntry = doGet(key, nextRetRev - 1, false);
+
+                                        evts.add(new EntryEvent(oldEntry, newEntry));
+                                    }
+                                }
+
+                                if (empty)
+                                    return null;
+
+                                if (evts.isEmpty())
+                                    continue;
+
+                                lastRetRev = nextRetRev;
+
+                                nextRetRev = -1;
+
+                                nativeIterator.close();
```

Comment:
should we reuse the iterator instead?

Reply:
Sadly, that is not possible. Newer data which was added during the lifetime of this iterator won't be "visible"

Reply:
that's why there's a [refresh](https://github.com/facebook/rocksdb/wiki/Iterator#resource-pinned-by-iterators-and-iterator-refreshing) method